modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -224,12 +289,98 @@ public static function validateForm(array &$element, FormStateInterface $form_st
       $form_state->setErrorByName($code_path, t('Coupon is invalid'));
       return;
     }
-    if (!$coupon->getPromotion()->applies($order)) {
-      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
-      return;
-    }
 
     $form_state->setValueForElement($element, $coupon);
   }
 
+  /**
+   * Adjustments table builder.
+   *
+   * @param array $element
+   *   The element.
+   * @param \Drupal\commerce_order\Entity\OrderInterface $order
+   *   The order.
+   * @param bool $display_actions
+   *   TRUE if actions displayed.
+   *
+   * @return array
+   *   The element array.
+   */
+  public static function buildAdjustmentsTable(array $element, OrderInterface $order, $display_actions = TRUE) {
+    $table = [
+      '#type' => 'table',
+      '#header' => [t('Label'), t('Amount')],
+      '#empty' => t('There are no special offers applied.'),
+    ];
+    if ($display_actions) {
+      $table['#header'][] = t('Remove');
+    }
+
+    $adjustments = $order->getAdjustments();
+    foreach ($order->getItems() as $orderItem) {
+      if ($item_adjustments = $orderItem->getAdjustments()) {
+        $adjustments = array_merge($adjustments, $item_adjustments);
+      }
+    }
+    $promotion_ids = array_map(function (Adjustment $adjustment) {
+      return $adjustment->getSourceId();
+    }, $adjustments);
+
+    /** @var \Drupal\commerce_promotion\Entity\CouponInterface[] $coupons */
+    $coupons = $order->get('coupons')->referencedEntities();
+    if (empty($coupons) || empty($adjustments)) {
+      return $table;
+    }
+
+    // Use special format for promotion with coupon.
+    /** @var \Drupal\commerce_promotion\Entity\CouponInterface $coupon */
+    foreach ($coupons as $index => $coupon) {
+      $adjustment_index = array_search($coupon->getPromotion()->id(), $promotion_ids);
+      $adjustment = $adjustments[$adjustment_index];
+
+      $label = t(':title (:code)', [
+        ':title' => $coupon->getPromotion()->getName(),
+        ':code' => $coupon->getCode(),
+      ]);
+      $table[$index]['label'] = [
+        '#type' => 'inline_template',
+        '#template' => '{{ label }}',
+        '#context' => [
+          'label' => $label,
+        ],
+      ];
+      $table[$index]['amount'] = [
+        '#type' => 'inline_template',
+        '#template' => '{{ price|commerce_price_format }}',
+        '#context' => [
+          'price' => $adjustment->getAmount(),
```

Comment:
The above logic seems insanely complex to just get the adjustment amount. Should review. Or open for a follow up.